File: plugins/vibeassist/skills/user-lens-map/scripts/routes_file_based.py
#!/usr/bin/env python3
"""Phase 1 route enumeration for FILE-BASED routers (TanStack Start, Expo
Router for React Native, Next-style dirs for simple cases). Generalized from the VA run's gen_routes.py: the
original hardcoded its layout list and redirect table by hand - here both are
DERIVED from the code, so the script works on repos it has never seen.

Rules carried over:
  - dot-nesting flattens: settings.account.tsx -> /settings/account
  - a trailing underscore un-nests: projects_.$projectId... escapes the
    /projects layout
  - pathless _segments (e.g. _authenticated) never appear in the URL
  - [.] protects a literal dot in a filename (sitemap[.]xml)
  Expo Router (React Native) conventions, auto-applied when the routes dir is
  app/ or expo-router is a dependency:
  - (group) segments are organizational only - stripped from the path
  - [param] -> :param, [...rest] -> a catch-all
  - _layout files are the layout itself, never a route
  - +not-found -> the not-found screen (flagged); name+api files -> machine
  - screens have no URLs: "reached from outside" means deep links and push
    notifications - note that in _stack.md
Derived (was hardcoded):
  - LAYOUT: a file whose segments are all pathless is the layout itself (drop);
    a file with dot-children siblings or a same-name sibling dir is a layout
    for them - it stays a route ONLY if it also renders its own component
  - REDIRECT: the file throws redirect({to}) in beforeLoad/loader ->
    audience "redirect" with redirect_to read from the code
  - MACHINE: the route declares no rendered component (structural test),
    with naming conventions as fallback
  - auth_required: true when wrapped by a pathless segment whose name
    contains "auth" (e.g. _authenticated); null otherwise - state the
    app's real guard in _stack.md

Usage:
    python scripts/routes_file_based.py --repo-root . -o map/_routes.json
    [--routes-dir src/routes]
"""
import argparse
import json
import pathlib
import re
import sys
from collections import Counter

try:
    sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception:
    pass

REDIRECT_TO = re.compile(r"""(?:throw\s+)?redirect\(\s*\{\s*to:\s*["'`]([^"'`]+)""")
HAS_COMPONENT = re.compile(r"component\s*:\s*\w+|export\s+(?:default\s+)?function\s+\w+\s*\([^)]*\)\s*\{[^}]*return\s*\(?\s*<|=>\s*\(?\s*<")
MACHINE_NAMES = re.compile(r"\.well-known|sitemap|robots|webhook|callback|\bfeed\b|healthz?", re.I)


def to_path(rel_after_routes: str) -> str:
    s = re.sub(r"\.(tsx|ts|jsx|js)$", "", rel_after_routes)
    s = s.replace("[.]", "\x00")  # protect literal dots
    parts = []
    for d in s.split("/"):
        parts.extend(d.split("."))
    parts = [p.replace("\x00", ".") for p in parts]
    segs = []
    for p in parts:
        if p.startswith("(") and p.endswith(")"):
            continue  # Expo group segment - organizational, not in the URL
        if p.startswith("[...") and p.endswith("]"):
            p = "*"  # catch-all
        elif p.startswith("[") and p.endswith("]"):
            p = ":" + p[1:-1]  # [id] -> :id
        if p.startswith("_") or p == "__root":
            continue  # pathless layout segment
        if p.endswith("_") and len(p) > 1:
            p = p[:-1]  # trailing underscore un-nests; URL segment keeps the name
        segs.append(p)
    if segs and segs[-1] == "index":
        segs.pop()
    return "/" + "/".join(segs) if segs else "/"


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--repo-root", default=".")
    ap.add_argument("--routes-dir", default="src/routes")
    ap.add_argument("-o", "--out", default="map/_routes.json")
    args = ap.parse_args()
    repo = pathlib.Path(args.repo_root).resolve()
    rdir = repo / args.routes_dir
    if not rdir.is_dir() and args.routes_dir == "src/routes" and (repo / "app").is_dir():
        rdir = repo / "app"  # Expo Router / app-dir convention
    if not rdir.is_dir():
        sys.exit(f"routes dir not found: {rdir} (pass --routes-dir)")

    files = sorted(p for p in rdir.rglob("*") if p.is_file() and p.suffix in (".tsx", ".ts", ".jsx", ".js"))
    stems = set()
    dirs = set()
    for f in files:
        rel = str(f.relative_to(rdir)).replace("\\", "/")
        stems.add(re.sub(r"\.(tsx|ts|jsx|js)$", "", rel))
    for d in rdir.rglob("*"):
        if d.is_dir():
            dirs.add(str(d.relative_to(rdir)).replace("\\", "/"))

    rows = []
    for f in files:
        rel_r = str(f.relative_to(rdir)).replace("\\", "/")
        rel = str(f.relative_to(repo)).replace("\\", "/")
        stem = re.sub(r"\.(tsx|ts|jsx|js)$", "", rel_r)
        segs = stem.replace("[.]", "\x00").split("/")
        flat = [s for part in segs for s in part.split(".")]

        # 1. pure layout: every segment is pathless -> not a route at all
        if all(s.startswith("_") or s == "__root" or (s.startswith("(") and s.endswith(")")) for s in flat):
            continue
        if flat[-1] == "_layout":
            continue  # Expo: the layout file for its folder, never a route itself

        text = f.read_text(encoding="utf-8", errors="replace")

        # 2. layout-with-children: dot-children or same-name dir exist
        has_children = any(s != stem and s.startswith(stem + ".") for s in stems) or (stem in dirs)
        renders_own = bool(HAS_COMPONENT.search(text))
        is_redirect_stub = bool(REDIRECT_TO.search(text)) and not renders_own
        if has_children and not renders_own and not is_redirect_stub:
            continue  # container only; its children are the routes
            # (a container that THROWS A REDIRECT is not a container - it is an
            #  old address kept working, and it must appear as audience "redirect")

        path = to_path(rel_r)
        not_found = flat[-1] == "+not-found"
        is_api = flat[-1].endswith("+api")
        rm = REDIRECT_TO.search(text)
        if rm and not renders_own:
            audience, redirect_to = "redirect", rm.group(1)
        elif is_api or (not renders_own and not not_found):
            # STRUCTURE FIRST, NAMES ONLY WHEN IT IS AMBIGUOUS — which is what
            # this skill's own Phase 1 says, and what the code was not doing. The
            # name test ran even on routes that plainly render a screen, so
            # `healthz?` matched "health" anywhere in a path and filed "Project
            # health check" (/projects/$id/tools/health) and "Setup & Health" —
            # two real pages sitting in the tools rail — as machine endpoints.
            # A route that renders a component is not ambiguous, so the naming
            # conventions only get a say when there is no component to judge.
            # (MACHINE_NAMES is kept below as documentation of the conventions;
            # the structural test now covers every case it used to catch.)
            audience, redirect_to = "machine", None
        else:
            audience, redirect_to = "user", None

        auth_wrappers = [s for s in flat if s.startswith("_") and "auth" in s.lower()]
        layout_chain = [s for s in flat[:-1] if s.startswith("_") or s == "__root"]
        row = {"path": path, "source_file": rel, "layout_chain": layout_chain,
               "auth_required": True if auth_wrappers else None, "audience": audience}
        if redirect_to:
            row["redirect_to"] = redirect_to
        if not_found:
            row["not_found"] = True
        rows.append(row)

    # ONE ROW PER PATH. A pathful layout and its index file resolve to the SAME
    # address: `admin.tsx` is the shell that renders around /admin/*, and
    # `admin.index.tsx` is what you actually get at /admin exactly. Both were
    # emitted — one "user" (the shell, which does render a component and so
    # survives the container test) and one "redirect" (the index, which is what
    # really happens) — so /admin, /settings and a project's config and tools
    # sections each appeared twice. On import that is a duplicate card per
    # section. Expo's _layout files are already dropped above; this is the
    # TanStack pathful-layout case.
    #
    # The index file wins: it IS the route at this address, and the same-named
    # file beside it is its layout.
    by_path: dict[str, dict] = {}
    for r in rows:
        if r["path"] not in by_path or re.search(r"\.index\.(tsx|ts|jsx|js)$", r["source_file"]):
            by_path[r["path"]] = r
    rows = list(by_path.values())

    rows.sort(key=lambda r: (r["audience"] != "user", r["path"]))
    out = pathlib.Path(args.out)
    out.parent.mkdir(parents=True, exist_ok=True)
    out.write_text(json.dumps(rows, indent=2), encoding="utf-8")
    print(dict(Counter(r["audience"] for r in rows)), "->", out)


if __name__ == "__main__":
    main()

File: plugins/vibeassist/skills/vibeassist-map/scripts/routes_file_based.mjs
#!/usr/bin/env node
// Faithful ESM port of routes_file_based.py — byte-for-byte parity with the
// Python CLI (argv, input reading, output file, stdout, exit codes).
// Translation, not redesign. See the .py for the domain documentation.
import fs from 'node:fs'
import path from 'node:path'
import process from 'node:process'
import { fileURLToPath } from 'node:url'
import { dumps } from './_pyjson.mjs'

// --- regexes translated from Python `re` ------------------------------------
const REDIRECT_TO = /(?:throw\s+)?redirect\(\s*\{\s*to:\s*["'`]([^"'`]+)/
const HAS_COMPONENT = /component\s*:\s*\w+|export\s+(?:default\s+)?function\s+\w+\s*\([^)]*\)\s*\{[^}]*return\s*\(?\s*<|=>\s*\(?\s*</
// MACHINE_NAMES is kept as documentation of the naming conventions; the
// structural test covers every case it used to catch. (Unused in Python too.)
// eslint-disable-next-line no-unused-vars
const MACHINE_NAMES = /\.well-known|sitemap|robots|webhook|callback|\bfeed\b|healthz?/i

const EXT_RE = /\.(tsx|ts|jsx|js)$/
const INDEX_RE = /\.index\.(tsx|ts|jsx|js)$/

// --- small helpers replicating Python-isms ----------------------------------

// Python str.replace replaces ALL occurrences (JS String.replace is first-only).
function replaceAll(s, find, repl) {
  return s.split(find).join(repl)
}

// pathlib.PurePosixPath(str(x)) normalization used by str(Path(...)):
// collapse repeated '/', drop '.' and empty segments, keep '..', keep a
// single leading '/'. Matches str(pathlib.Path("./a//b/")) -> "a/b".
function posixPathStr(p) {
  if (p === '') return '.'
  const isAbs = p.startsWith('/')
  // pathlib treats a leading '//' specially, but that never occurs here.
  const parts = p.split('/').filter((seg) => seg !== '' && seg !== '.')
  const joined = parts.join('/')
  if (isAbs) return '/' + joined
  return joined === '' ? '.' : joined
}

// pathlib PurePath.suffix
function suffixOf(name) {
  const i = name.lastIndexOf('.')
  if (i > 0 && i < name.length - 1) return name.slice(i)
  return ''
}

// Python repr() of a str (for dict-repr keys). Audience values are simple, but
// keep it faithful for safety.
function pyStrRepr(s) {
  const hasSingle = s.includes("'")
  const hasDouble = s.includes('"')
  const quote = hasSingle && !hasDouble ? '"' : "'"
  let out = quote
  for (const ch of s) {
    const cp = ch.codePointAt(0)
    if (ch === '\\') out += '\\\\'
    else if (ch === quote) out += '\\' + quote
    else if (ch === '\n') out += '\\n'
    else if (ch === '\r') out += '\\r'
    else if (ch === '\t') out += '\\t'
    else if (cp < 0x20 || cp === 0x7f) out += '\\x' + cp.toString(16).padStart(2, '0')
    else out += ch
  }
  return out + quote
}

// repr of dict(Counter(...)) — insertion order = first-encounter order.
function pyDictRepr(entries) {
  const inner = entries.map(([k, v]) => `${pyStrRepr(k)}: ${v}`).join(', ')
  return '{' + inner + '}'
}

// --- to_path (translated verbatim) ------------------------------------------
function toPath(relAfterRoutes) {
  let s = relAfterRoutes.replace(EXT_RE, '')
  s = replaceAll(s, '[.]', '\x00') // protect literal dots
  let parts = []
  for (const d of s.split('/')) {
    for (const seg of d.split('.')) parts.push(seg)
  }
  parts = parts.map((p) => replaceAll(p, '\x00', '.'))
  const segs = []
  for (let p of parts) {
    if (p.startsWith('(') && p.endsWith(')')) continue // Expo group segment
    if (p.startsWith('[...') && p.endsWith(']')) {
      p = '*' // catch-all
    } else if (p.startsWith('[') && p.endsWith(']')) {
      p = ':' + p.slice(1, -1) // [id] -> :id
    }
    if (p.startsWith('_') || p === '__root') continue // pathless layout segment
    if (p.endsWith('_') && p.length > 1) p = p.slice(0, -1) // trailing underscore un-nests
    segs.push(p)
  }
  if (segs.length && segs[segs.length - 1] === 'index') segs.pop()
  return segs.length ? '/' + segs.join('/') : '/'
}

// --- filesystem: replicate rdir.rglob("*") -> is_file/is_dir ----------------
// pathlib rglob includes dotfiles and hidden dirs; sorted() over Paths sorts by
// the full path string (PurePosixPath comparison). We gather all files/dirs and
// sort files by absolute path string to match.
function walk(root) {
  const files = []
  const dirs = []
  function recurse(dir) {
    let entries
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true })
    } catch {
      return
    }
    for (const ent of entries) {
      const full = path.join(dir, ent.name)
      let isDir = ent.isDirectory()
      let isFile = ent.isFile()
      if (ent.isSymbolicLink()) {
        try {
          const st = fs.statSync(full)
          isDir = st.isDirectory()
          isFile = st.isFile()
        } catch {
          isDir = false
          isFile = false
        }
      }
      if (isDir) {
        dirs.push(full)
        recurse(full)
      } else if (isFile) {
        files.push(full)
      }
    }
  }
  recurse(root)
  return { files, dirs }
}

function isDir(p) {
  try {
    return fs.statSync(p).isDirectory()
  } catch {
    return false
  }
}

// --- arg parsing (argparse subset) ------------------------------------------
function parseArgs(argv) {
  const args = { repo_root: '.', routes_dir: 'src/routes', out: 'map/_routes.json' }
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i]
    const takeVal = (inline) => (inline !== undefined ? inline : argv[++i])
    if (a === '--repo-root' || a.startsWith('--repo-root=')) {
      args.repo_root = takeVal(a.includes('=') ? a.slice(a.indexOf('=') + 1) : undefined)
    } else if (a === '--routes-dir' || a.startsWith('--routes-dir=')) {
      args.routes_dir = takeVal(a.includes('=') ? a.slice(a.indexOf('=') + 1) : undefined)
    } else if (a === '-o' || a === '--out' || a.startsWith('--out=')) {
      args.out = takeVal(a.includes('=') ? a.slice(a.indexOf('=') + 1) : undefined)
    } else if (a.startsWith('-o') && a.length > 2 && !a.startsWith('--')) {
      args.out = a.slice(2)
    }
  }
  return args
}

function main() {
  const args = parseArgs(process.argv.slice(2))
  const repo = path.resolve(args.repo_root)
  let rdir = path.resolve(repo, args.routes_dir)
  if (!isDir(rdir) && args.routes_dir === 'src/routes' && isDir(path.resolve(repo, 'app'))) {
    rdir = path.resolve(repo, 'app') // Expo Router / app-dir convention
  }
  if (!isDir(rdir)) {
    process.stderr.write(`routes dir not found: ${rdir} (pass --routes-dir)\n`)
    process.exit(1)
  }

  const walked = walk(rdir)
  const files = walked.files
    .filter((p) => ['.tsx', '.ts', '.jsx', '.js'].includes(suffixOf(path.basename(p))))
    .sort((a, b) => (a < b ? -1 : a > b ? 1 : 0))

  const relTo = (base, p) => path.relative(base, p).split(path.sep).join('/')

  const stems = new Set()
  const dirs = new Set()
  for (const f of files) {
    const rel = relTo(rdir, f)
    stems.add(rel.replace(EXT_RE, ''))
  }
  for (const d of walked.dirs) {
    dirs.add(relTo(rdir, d))
  }

  const rows = []
  for (const f of files) {
    const rel_r = relTo(rdir, f)
    const rel = relTo(repo, f)
    const stem = rel_r.replace(EXT_RE, '')
    const segs = replaceAll(stem, '[.]', '\x00').split('/')
    const flat = []
    for (const part of segs) for (const s of part.split('.')) flat.push(s)

    // 1. pure layout: every segment pathless -> not a route at all
    if (
      flat.every(
        (s) => s.startsWith('_') || s === '__root' || (s.startsWith('(') && s.endsWith(')'))
      )
    ) {
      continue
    }
    if (flat[flat.length - 1] === '_layout') continue // Expo layout file, never a route

    const text = fs.readFileSync(f, 'utf-8')

    // 2. layout-with-children
    const has_children =
      Array.from(stems).some((s) => s !== stem && s.startsWith(stem + '.')) || dirs.has(stem)
    const renders_own = HAS_COMPONENT.test(text)
    const is_redirect_stub = REDIRECT_TO.test(text) && !renders_own
    if (has_children && !renders_own && !is_redirect_stub) continue // container only

    const p = toPath(rel_r)
    const not_found = flat[flat.length - 1] === '+not-found'
    const is_api = flat[flat.length - 1].endsWith('+api')
    const rm = REDIRECT_TO.exec(text)
    let audience
    let redirect_to
    if (rm && !renders_own) {
      audience = 'redirect'
      redirect_to = rm[1]
    } else if (is_api || (!renders_own && !not_found)) {
      audience = 'machine'
      redirect_to = null
    } else {
      audience = 'user'
      redirect_to = null
    }

    const auth_wrappers = flat.filter((s) => s.startsWith('_') && s.toLowerCase().includes('auth'))
    const layout_chain = flat.slice(0, -1).filter((s) => s.startsWith('_') || s === '__root')
    const row = {
      path: p,
      source_file: rel,
      layout_chain: layout_chain,
      auth_required: auth_wrappers.length ? true : null,
      audience: audience,
    }
    if (redirect_to) row.redirect_to = redirect_to
    if (not_found) row.not_found = true
    rows.push(row)
  }

  // ONE ROW PER PATH — index file wins.
  const by_path = new Map()
  for (const r of rows) {
    if (!by_path.has(r.path) || INDEX_RE.test(r.source_file)) {
      by_path.set(r.path, r)
    }
  }
  let finalRows = Array.from(by_path.values())

  // rows.sort(key=lambda r: (r["audience"] != "user", r["path"])) — stable.
  finalRows = finalRows
    .map((r, i) => [r, i])
    .sort((A, B) => {
      const a = A[0]
      const b = B[0]
      const ka = a.audience !== 'user' ? 1 : 0
      const kb = b.audience !== 'user' ? 1 : 0
      if (ka !== kb) return ka - kb
      if (a.path < b.path) return -1
      if (a.path > b.path) return 1
      return A[1] - B[1] // stable tie-break
    })
    .map((x) => x[0])

  const out = args.out
  const outDir = path.dirname(out) || '.'
  fs.mkdirSync(outDir, { recursive: true })
  fs.writeFileSync(out, dumps(finalRows, 2))

  // print(dict(Counter(r["audience"] for r in rows)), "->", out)
  const counter = new Map()
  for (const r of finalRows) counter.set(r.audience, (counter.get(r.audience) || 0) + 1)
  const dictRepr = pyDictRepr(Array.from(counter.entries()))
  process.stdout.write(`${dictRepr} -> ${posixPathStr(out)}\n`)
}

// Guard: run only when executed directly.
const isMain =
  process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)
if (isMain) {
  main()
}

export { toPath, main }
